notes-0422-normalization.tex

```diff
@@ -0,0 +1,94 @@
+
+\section{Optimizations}
+
+\subsection{Function Normalization}
+If a straightforward normalization strategy is used, some parts of an expression might be normalized multiple times. For example, to normalize the expression $\app{\parens{\lam{x}{M}}}{N}$, we first normalize $M$ to $M'$, and $N$ to $N'$. After that, we will normalize $M'\subst{N'}{x}$. In this case, the expression $M$ is normalized twice.
+
+One possible optimization is to delay the normalization of $M$. When we evaluate a term to the form $\lam{x}{M}$, the inner part is not normalized immediately, but it is rather put into a separate domain. The operation from the terms to the domain is called ``evaluate'', and the other direction is called ``reify''. These two operations are intertwined, because when $\lam{x}{M}$ is reified, $M$ should be further evaluated.
+
+\subsection{De Bruijn Index and De Bruijn Level}
+The term equality we were talking about is under $\alpha$-equivalence, that is, two terms are equal if they can be syntactically equal after appropriate renamings. For example, $\lam{x}{\app{x}{\parens{\lam{y}{\app{x}{y}}}}}$ and $\lam{y}{\app{y}{\parens{\lam{x}{\app{y}{x}}}}}$ are equal. But in practice, it is surprisingly tricky to handle the renaming and substitution correctly.
+
+One way to handle the variable names is using De Bruijn index, which uses natural numbers to name variables. A variable name is represented by the number of ``$\lambda$''s between its occurrence and its binding site. For example, both $\lam{x}{\app{x}{\parens{\lam{y}{\app{x}{y}}}}}$ and $\lam{y}{\app{y}{\parens{\lam{x}{\app{y}{x}}}}}$ are represented by the same term $\lam{}{\app{0}{\parens{\lam{}{\app{1}{0}}}}}$. Notice that a variable can have different names at different places, like the first 0 and 1 both refer to the variable $x$ in
+$\lam{x}{\app{x}{\parens{\lam{y}{\app{x}{y}}}}}$.
+
+Using De Bruijn index, the equality of normal forms is just syntactical equality. However, during the evaluation inside the functions, we still have to handle variable renamings. Whenever we move into another layer of $\lambda$, we need to increment all the variable names in the context by 1. This can be avoided by using De Bruijn level in the domain.
+
+De Bruijn level is roughly the opposite of De Bruijn index. The variable name is represented by the number of ``$\lambda$''s before its binding site. So the term $\lam{x}{\app{x}{\parens{\lam{y}{\app{x}{y}}}}}$ is represented by $\lam{}{\app{0}{\parens{\lam{}{\app{0}{1}}}}}$. Using De Bruijn level, the variable name is like a constant value, so we don't have to shift the variable names in the context. It is also easy to convert between De Bruijn index and level. If a variable appears in the $n$th $\lambda$, then the sum of the index and the level should be $(n-1)$.
+
+Therefore, the variable naming problem can be solved by using De Bruijn index for the terms, and using De Bruijn level for the domain.
```

Comment:
The reason that we can avoid shifting is highly dependent on how a typical normalization-by-evaluation algorithm works. It's not because De Bruijn levels are more powerful.